core/modules/file/file.pages.inc

```diff
@@ -339,31 +339,48 @@ function file_add_form_submit($form, &$form_state) {
       if (file_uri_scheme($file->uri) != $form_state['storage']['scheme']) {
         $file_destination = $form_state['storage']['scheme'] . '://' . file_uri_target($file->uri);
         $file_destination = file_stream_wrapper_uri_normalize($file_destination);
+        $file_destination_dirname = backdrop_dirname($file_destination);
+        // Create the directory in case it doesn't exist.
+        file_prepare_directory($file_destination_dirname, FILE_CREATE_DIRECTORY);
         if ($moved_file = file_move($file, $file_destination, FILE_EXISTS_RENAME)) {
           // Only re-assign the file object if file_move() did not fail.
           $file = $moved_file;
         }
       }
-      $file->type = $form_state['storage']['type'];
-      $file->display = TRUE;
-
-      // Change the file from temporary to permanent.
-      $file->status = FILE_STATUS_PERMANENT;
-
-      // Save the form fields.
-      // Keep in mind that the values for the Field API fields must be in
-      // $form_state['values'] and not in ['storage']. This is true as long as
-      // the fields are on the last page of the multi step form.
-      entity_form_submit_build_entity('file', $file, $form, $form_state);
-
-      file_save($file);
-      // Record that the module (in this case, the File module) is using the
-      // file. Without the call to file_usage_add, file_managed_file_validate()
-      // produces an error upon saving the form, saying that the uploaded file
-      // may not be referenced.
-      file_usage_add($file, 'file', 'file', $file->fid);
-      $form_state['file'] = $file;
-      backdrop_set_message(t('@type %name was uploaded.', array('@type' => file_type_get_name($file), '%name' => $file->filename)));
+
+      $substitutions = array(
+        '@type' => file_type_get_name($file),
+        '%name' => $file->filename,
+      );
+
+      // Only save the file if the move completed.
+      if ($moved_file){
```

Comment:
$moved_file can be undefined.

See line 345, where it's defined inside an if clause.